core/iwasm/libraries/debug-engine/gdbserver.c

```diff
@@ -0,0 +1,177 @@
+    DEL_HANDLER('z', handle_remove_break),
+    DEL_HANDLER('c', handle_continue_request),
+    DEL_HANDLER('k', handle_kill_request),
+    DEL_HANDLER('_', handle____request),
+};
+
+WASMGDBServer *
+wasm_launch_gdbserver(char *host, int port)
+{
+    int listen_fd = -1;
+    const int one = 1;
+    struct sockaddr_in addr;
+    int ret;
+    int sockt_fd = 0;
+
+    WASMGDBServer *server;
+
+    if (!(server = wasm_runtime_malloc(sizeof(WASMGDBServer)))) {
+        LOG_ERROR("wasm gdb server error: failed to allocate memory");
+        return NULL;
+    }
+
+    listen_fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+    if (listen_fd < 0) {
+        LOG_ERROR("wasm gdb server error: socket() failed");
+        goto fail;
+    }
+
+    ret = fcntl(listen_fd, F_SETFD, FD_CLOEXEC);
+    if(ret < 0) {
+        LOG_ERROR("wasm gdb server error: fcntl() failed on setting FD_CLOEXEC");
+        goto fail;
+    }
+
+    ret = setsockopt(listen_fd, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
+    if (ret < 0) {
+        LOG_ERROR("wasm gdb server error: setsockopt() failed");
+        goto fail;
+    }
+
+    LOG_VERBOSE("Listening on %s:%d\n", host, port);
+    addr.sin_family = AF_INET;
+    addr.sin_addr.s_addr = inet_addr(host);
+    addr.sin_port = htons(port);
+
+    ret = bind(listen_fd, (struct sockaddr *)&addr, sizeof(addr));
+    if (ret < 0) {
+        LOG_ERROR("wasm gdb server error: bind() failed");
+        goto fail;
+    }
+
+    ret = listen(listen_fd, 1);
+    if (ret < 0) {
+        LOG_ERROR("wasm gdb server error: listen() failed");
+        goto fail;
+    }
+
+    server->listen_fd = listen_fd;
+
+    sockt_fd = accept(listen_fd, NULL, NULL);
+    if (sockt_fd < 0) {
+        LOG_ERROR("wasm gdb server error: accept() failed");
+        goto fail;
+    }
+    LOG_VERBOSE("accept gdb client");
+    server->socket_fd = sockt_fd;
+    server->noack = false;
+    return server;
+
+fail:
+    if (listen_fd > 0) {
+        shutdown(listen_fd, SHUT_RDWR);
+        close(listen_fd);
+    }
+    if (server)
+        wasm_runtime_free(server);
+    return NULL;
+}
+
+void
+wasm_close_gdbserver(WASMGDBServer *server)
+{
+    if (server->socket_fd > 0) {
+        shutdown(server->socket_fd, SHUT_RDWR);
+        close(server->socket_fd);
+    }
+    if (server->listen_fd > 0) {
+        shutdown(server->listen_fd, SHUT_RDWR);
+        close(server->listen_fd);
+    }
+}
+
+static inline void
+handler_packet(WASMGDBServer *server, char request, char *payload)
+{
+    if (packet_handler_table[(int)request].handler != NULL)
+        packet_handler_table[(int)request].handler(server, payload);
+}
+
+static void
+process_packet(WASMGDBServer *server)
+{
+    uint8_t *inbuf = server->pkt.buf;
+    int inbuf_size = server->pkt.end;
+    uint8_t *packetend_ptr = (uint8_t *)memchr(inbuf, '#', inbuf_size);
+    int packetend = packetend_ptr - inbuf;
+    bh_assert('$' == inbuf[0]);
+    char request = inbuf[1];
```

Comment:
in some conditions `pkt.end` can be 1. which means the `inbuf[1]` and `inbuf[2]` will read invalid memory, which won't necessarily cause a crash but is still a serious issue.

Reply:
if there's a hard-coded `inbuf[2]` better put an assert but IMO it's the protocol handling code needs some more work.

Reply:
btw, the name `pkt.end` is confusing. considering there's a `pkt.buf`, the `pkt.end` sounds like a pointer to the end. Should've used the name `size` instead.